pageserver/src/tenant/timeline/walreceiver/connection_manager.rs

```diff
@@ -214,6 +228,52 @@ pub(super) async fn connection_manager_loop_step(
                     }
                 }
             } => debug!("Waking up for the next retry after waiting for {time_until_next_retry:?}"),
+
+            Some(()) = async {
+                // Calculating time needed to wait until sending the next discovery request.
+                // Current implementation is conservative and sends discovery requests only when there are no candidates.
+
+                if any_activity {
+                    // No need to send discovery requests if there is an active connection or candidates.
+                    return None;
+                }
+
+                // Waiting for an active wait_lsn request.
+                while wait_lsn_status.borrow().is_none() {
+                    wait_lsn_status.changed().await.ok();
+                }
+
+                // All preconditions met, preparing to send a discovery request.
+                let now = Utc::now().naive_utc();
+                let next_discovery_ts = last_discovery_ts
+                    .checked_add_signed(discovery_request_interval)
+                    .unwrap_or(now);
+
+                if next_discovery_ts > now {
+                    // Prevent sending discovery requests too frequently.
+                    tokio::time::sleep((next_discovery_ts - now).to_std().unwrap_or_default()).await;
+                }
+
+                let tenant_timeline_id = Some(ProtoTenantTimelineId {
+                    tenant_id: id.tenant_id.as_ref().to_owned(),
+                    timeline_id: id.timeline_id.as_ref().to_owned(),
+                });
+                let request = SafekeeperDiscoveryRequest { tenant_timeline_id };
+                let msg = TypedMessage {
+                    r#type: MessageType::SafekeeperDiscoveryRequest as i32,
+                    safekeeper_timeline_info: None,
+                    safekeeper_discovery_request: Some(request),
+                    safekeeper_discovery_response: None,
+                    };
+
+                last_discovery_ts = Utc::now().naive_utc();
+                debug!("No active connection and no candidates, sending discovery request to the broker");
+
+                // This is a fire-and-forget request, we don't care about the response
+                let _ = broker_client.publish_one(msg).await;
+                debug!("Discovery request sent to the broker");
+                None
```

Comment:
cancellation safety: If we get cancelled at this .await point, we have already updated `last_discovery_ts` but haven't necessarily completed the transmission of the message.

Reply:
Yes, it is expected. I considered doing `tokio::spawn`, but cancellation seems more appropriate here.

I'll add a comment about it.

Reply:
I see you added a comment wrt cancellation safety.

But, isn't the point of `last_discover_ts` to track the last successful discovery request?

If we get cancelled at the `publish_one()` await point, we haven't yet sent the request.

So, I would argue that `last_discovery_ts` must be updated _after_ `publish_one` returns.

Reply:
I'm afraid that if `last_discovery_ts` will be updated after `publish_one`, we can send too many requests and overload something.

Currently discovery requests are supposed to be just fallback option, and IMO sending less requests is better option than sending more requests.